packages/scripts/src/assessCases.ts

```diff
@@ -0,0 +1,228 @@
+              },
+            ];
+          }
+        )
+      );
+      return {
+        embeddings: variantEmbeddings,
+        prompt,
+        relevance,
+      };
+    }
+  );
+};
+
+export const assessRelevance = async ({
+  prompt,
+  embedders,
+  expected,
+  generate,
+}: {
+  prompt: string;
+  expected: string;
+  embedders: Embedder[];
+  generate: SimpleTextGenerator;
+}): Promise<Relevance> => {
+  const shortName = makeShortName(prompt);
+  console.log(`Calculating embeddings for '${shortName}'...`);
+
+  const promptEmbeddings = await calculateEmbeddings({
+    text: prompt,
+    embedders,
+  });
+
+  console.log(`Generating variants for '${shortName}'...`);
+  const variants = await generatePromptsFromExpectedAnswer({
+    embedders,
+    expected,
+    generate,
+    howMany: 3,
+  });
+
+  const variantCount = Object.values(variants).length;
+  if (variantCount === 0) {
+    throw new Error(`Unexpectedly without variants for ${shortName}!`);
+  }
+
+  const scoredVariants = scoreVariants({
+    original: { prompt, embeddings: promptEmbeddings },
+    variants,
+  });
+  assert(variantCount === Object.values(scoredVariants).length);
+
+  console.log(
+    `- Expected: "${expected}"
+- Original: "${prompt}"
+- Generated variants:
+${scoredVariants.map(({ prompt }) => `  - "${prompt}"`).join("\n")}`
+  );
+
+  const summedMetrics = scoredVariants.reduce(
+    (outer, { relevance }): RelevanceMetrics => {
+      const relevanceValues = Object.values(relevance);
+      const modelCount = relevanceValues.length;
+      assert(modelCount > 0);
+      // Sum metrics across models
+      const crossModel = relevanceValues.reduce(
+        (acc, { cos_similarity, norm_sq_mag_diff }) => ({
+          cos_similarity: acc.cos_similarity + cos_similarity,
+          norm_sq_mag_diff: acc.norm_sq_mag_diff + norm_sq_mag_diff,
+        }),
+        { cos_similarity: 0, norm_sq_mag_diff: 0 }
+      );
+
+      // Accumulate averages across models
+      return {
+        cos_similarity:
+          outer.cos_similarity + crossModel.cos_similarity / modelCount,
+        norm_sq_mag_diff:
+          outer.norm_sq_mag_diff + crossModel.norm_sq_mag_diff / modelCount,
+      };
+    },
+    { cos_similarity: 0, norm_sq_mag_diff: 0 }
+  );
+  const averages: RelevanceMetrics = {
+    cos_similarity: summedMetrics.cos_similarity / variantCount,
+    norm_sq_mag_diff: summedMetrics.norm_sq_mag_diff / variantCount,
+  };
+
+  console.log(`Average score: ${JSON.stringify(averages)}`);
+  return {
+    prompt_embeddings: promptEmbeddings,
+    generated_prompts: scoredVariants,
+    averages,
+  };
+};
+
+export const rateWithLlm = async ({
+  prompt,
+  expected,
+  generate,
+}: {
+  prompt: string;
+  expected: string;
+  generate: SimpleTextGenerator;
+}): Promise<LlmAsJudgment | undefined> => {
+  const shortName = makeShortName(prompt);
+  console.log(`Rating '${shortName}' with LLM...`);
+
+  const [response] = await generate({
```

Comment:
rather than raw text gen and telling it to do json, models support structured output generation, which'd be perfect for this use case.

i recommend using the AI SDK [generateObject()](https://ai-sdk.dev/docs/reference/ai-sdk-core/generate-object) function. takes schema from zod type and handles output parsing.